book/src/design.md

```diff
@@ -1 +1,18 @@
 # Design
+
+## General design notes
+
+### Typed variables vs byte encodings
+
+For Sapling, we have encountered multiple places where the specification uses typed
+variables to define the consensus rules, but the C++ implementation in zcashd relied on
+byte encodings to implement them. This resulted in subtly-different consensus rules being
+deployed than were intended, for example where a particular type was not round-trip
+encodable.
+
+In Pollard, we avoid this by defining the consensus rules in terms of the byte encodings
+of all variables, and being explicit about any types that are not round-trip encodable.
+This makes consensus compatibility between strongly-typed implementations (such as this
+crate) and byte-oriented implementations easier to achieve.
+
+- TODO: Do we want to take this approach, or try again for "everything non-canonical"?
```

Comment:
I disagree with this approach.

In Sapling we failed to implement some of what the spec said &mdash; for example, that abst<sub>J</sub> should return ⊥ for any encoding not in the image of repr<sub>J</sub>. I don't think that the approach of defining an abstract protocol involving high-level types was the issue here.

There were three contributing factors:
* we just didn't check the implementation against the spec thoroughly enough;
* the spec included non-constructive definitions (such as the one above);
* the consensus rules weren't named in a way that could be linked to their implementations.

We can fix these problems without losing the ability to define consensus rules on high-level types.

Reply:
I tend to agree here. By defining consensus rules relative to high-level types, those types represent explicit sets of values, rather than encodings of those values. I think this is *more* compatible with strongly-typed implementations than defining rules in terms of wire encodings would be. There are two "layers" of consensus, as I see it: representational consensus (in terms of the ability to parse from a particular byte representation) and semantic consensus, which is better addressed by defining directly the membership of sets of valid values (the high-level types) and then the consensus rules that relate the interaction of those types.

Reply:
I'm now convinced that we can get the best of both worlds: explicitly define the conversions between transaction fields represented as byte sequences, and also be able to analyse the protocol in terms of high-level types. So I consider this resolved.